### PR TITLE
Bolt: Add E4S testsuite stand alone test

### PR DESCRIPTION
**Pull request by RikkiButler20**
This PR represents a preliminary stand alone test for the `bolt` package and is intended to serve the following purposes:

1. Leverage the current E4S test suite for the software
2. Demonstrate to package maintainers how to start a Spack stand alone test.

### CODE REVIEW

**Comment by shintaro-iwasaki** [comment]
Thanks! It looks good to me.

File: var/spack/repos/builtin/packages/bolt/package.py [hunk 1]
```diff
@@ -40,3 +42,30 @@ def cmake_args(self):
...
+    @run_after('install')
+    def cache_test_sources(self):
+        """Copy the example source files after the package is installed to an
+        install test subdirectory for use during `spack test run`."""
+        self.cache_extra_test_sources(['examples'])
+
+    def run_sample_nested_example(self):
+        """Run stand alone test: sample_nested"""
+
+        test_dir = join_path(self.install_test_root, 'examples')
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/bolt/package.py]: Please use the test stage directory copy for these tests:

```suggestion
        test_dir = join_path(self.test_suite.current_test_cache_dir, 'examples')
```

**Review by tldahlgren — CHANGES_REQUESTED** [review]

**Comment by tldahlgren** [comment]
Please also resolve the style issues:

```
  var/spack/repos/builtin/packages/bolt/package.py:64: [E128] continuation line under-indented for visual indent
20
  var/spack/repos/builtin/packages/bolt/package.py:66: [E128] continuation line under-indented for visual indent
21
  var/spack/repos/builtin/packages/bolt/package.py:67: [E128] continuation line under-indented for visual indent
```

File: var/spack/repos/builtin/packages/bolt/package.py [hunk 2]
```diff
@@ -27,6 +29,8 @@ class Bolt(CMakePackage):
     version("1.0.1", sha256="769e30dfc4042cee7ebbdadd23cf08796c03bcd8b335f516dc8cbc3f8adfa597")
     version("1.0", sha256="1c0d2f75597485ca36335d313a73736594e75c8a36123c5a6f54d01b5ba5c384")
 
+    test_requires_compiler = True
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/bolt/package.py:32]: Great catch!

File: var/spack/repos/builtin/packages/bolt/package.py [hunk 3]
```diff
@@ -40,3 +44,35 @@ def cmake_args(self):
...
+
+        test_dir = join_path(self.test_suite.current_test_cache_dir, 'examples')
+
+        if not os.path.exists(test_dir):
+            print('Skipping bolt test')
+            return
+
+        exe = 'sample_nested'
+
+        self.run_test('gcc',
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/bolt/package.py:67]: This worked for me as-is when I tested it, thanks!  

My concern in general is that the package is supposed to be buildable with a variety of compilers and this test case only checks one.  It is preferable that it use `self.compiler.cc` so the test builds with the same compiler used when the code is built.  Unfortunately, that simple change didn't work for me.  But the problem may be part of another issue.

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/bolt/package.py:67]:
```suggestion
        # TODO: Either change to use self.compiler.cc (so using the build-time compiler)
        #  or add test parts that compile with the different supported compilers.
        self.run_test('gcc',
```

**Review by tldahlgren — APPROVED** [review]

**Pull request merged**